File: wear/src/main/java/com/pdyjak/powerampwear/music_browser/folders/FoldersBrowserFragment.java
package com.pdyjak.powerampwear.music_browser.folders;

import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.android.gms.wearable.MessageEvent;
import com.pdyjak.powerampwear.MessageExchangeHelper;
import com.pdyjak.powerampwear.MessageListener;
import com.pdyjak.powerampwear.R;
import com.pdyjak.powerampwear.music_browser.BrowserFragmentBase;
import com.pdyjak.powerampwear.music_browser.Clickable;
import com.pdyjak.powerampwear.music_browser.ItemViewHolder;
import com.pdyjak.powerampwear.music_browser.MusicLibraryNavigator;
import com.pdyjak.powerampwear.music_browser.ViewHolderFactory;
import com.pdyjak.powerampwearcommon.requests.RequestsPaths;
import com.pdyjak.powerampwearcommon.responses.Folder;
import com.pdyjak.powerampwearcommon.responses.FoldersListResponse;

import java.util.ArrayList;
import java.util.List;

public class FoldersBrowserFragment extends BrowserFragmentBase implements MessageListener {

    private static class FoldersViewHolderFactory implements ViewHolderFactory {
        @NonNull
        @Override
        public ItemViewHolder createViewHolder(ViewGroup parent) {
            View view = LayoutInflater.from(parent.getContext()).inflate(
                    R.layout.basic_two_line_item, parent, false
            );
            return new FolderViewHolder(view);
        }
    }

    @NonNull
    @Override
    protected ViewHolderFactory createViewHolderFactory() {
        return new FoldersViewHolderFactory();
    }

    @Override
    protected boolean tryRestoreCachedItems() {
        FoldersListResponse cached = getMusicLibraryCache().getFoldersList();
        if (cached != null) {
            setItems(transform(cached.getFoldersList()));
            return true;
        }
        return false;
    }

    @Override
    protected void fetchItems() {
        MessageExchangeHelper helper = getMessageExchangeHelper();
        helper.addMessageListenerWeakly(this);
        helper.sendRequest(RequestsPaths.GET_FOLDERS);
    }

    @Override
    public void onPause() {
        super.onPause();
        getMessageExchangeHelper().removeMessageListener(this);
    }

    @Override
    public void onMessageReceived(@NonNull MessageEvent messageEvent) {
        if (FoldersListResponse.PATH.equals(messageEvent.getPath())) {
            getMessageExchangeHelper().removeMessageListener(this);
            byte[] bytes = messageEvent.getData();
            if (bytes == null) return;
            FoldersListResponse response = FoldersListResponse.fromBytes(bytes);
            if (response != null) {
                getMusicLibraryCache().update(response);
                setItems(transform(response.getFoldersList()));
            }
        }
    }

    private List<Clickable> transform(@NonNull List<Folder> folders) {
        MusicLibraryNavigator navigator = getMusicLibraryNavigator();
        List<Clickable> items = new ArrayList<>();
        for (Folder folder : folders) {
            items.add(new FolderItem(navigator, folder.id, folder.name, folder.parentName));
        }
        return items;
    }
}

File: wear/src/main/java/com/pdyjak/powerampwear/music_browser/folders/FoldersBrowserFragment.kt
package com.pdyjak.powerampwear.music_browser.folders

import android.view.LayoutInflater
import android.view.ViewGroup

import com.google.android.gms.wearable.MessageEvent
import com.pdyjak.powerampwear.MessageListener
import com.pdyjak.powerampwear.R
import com.pdyjak.powerampwear.messageExchangeHelper
import com.pdyjak.powerampwear.musicLibraryCache
import com.pdyjak.powerampwear.musicLibraryNavigator
import com.pdyjak.powerampwear.music_browser.BrowserFragmentBase
import com.pdyjak.powerampwear.music_browser.Clickable
import com.pdyjak.powerampwear.music_browser.ItemViewHolder
import com.pdyjak.powerampwear.music_browser.ViewHolderFactory
import com.pdyjak.powerampwearcommon.requests.RequestsPaths
import com.pdyjak.powerampwearcommon.responses.Folder
import com.pdyjak.powerampwearcommon.responses.FoldersListResponse

class FoldersBrowserFragment : BrowserFragmentBase(), MessageListener {

    private class FoldersViewHolderFactory : ViewHolderFactory {
        override fun createViewHolder(parent: ViewGroup): ItemViewHolder {
            val view = LayoutInflater.from(parent.context).inflate(
                    R.layout.basic_two_line_item, parent, false
            )
            return FolderViewHolder(view)
        }
    }

    override fun createViewHolderFactory(): ViewHolderFactory {
        return FoldersViewHolderFactory()
    }

    override fun tryRestoreCachedItems(): Boolean {
        val cached = activity.musicLibraryCache.foldersList
        setItems(transform(cached?.foldersList ?: return false))
        return true
    }

    override fun fetchItems() {
        val helper = activity.messageExchangeHelper
        helper.addMessageListenerWeakly(this)
        helper.sendRequest(RequestsPaths.GET_FOLDERS)
    }

    override fun onPause() {
        super.onPause()
        activity.messageExchangeHelper.removeMessageListener(this)
    }

    override fun onMessageReceived(messageEvent: MessageEvent) {
        if (FoldersListResponse.PATH == messageEvent.path) {
            activity.messageExchangeHelper.removeMessageListener(this)
            val bytes = messageEvent.data ?: return
            val response = FoldersListResponse.fromBytes(bytes)
            if (response != null) {
                activity.musicLibraryCache.update(response)
                setItems(transform(response.foldersList))
            }
        }
    }

    private fun transform(folders: List<Folder>): List<Clickable> {
        return folders.map {
            FolderItem(activity.musicLibraryNavigator, it.id, it.name, it.parentName)
        }
    }
}
